langlib/lang.array/src/main/java/org/ballerinalang/langlib/array/Next.java

```diff
@@ -49,16 +59,34 @@
     //TODO: refactor hard coded values
     public static Object next(Strand strand, ObjectValue m) {
         IteratorValue arrIterator = (IteratorValue) m.getNativeData("&iterator&");
-
+        ArrayValue arr = (ArrayValue) m.get("m");
         if (arrIterator == null) {
-            arrIterator = ((ArrayValue) m.get("m")).getIterator();
+            arrIterator = arr.getIterator();
             m.addNativeData("&iterator&", arrIterator);
         }
 
         if (arrIterator.hasNext()) {
             Object element =  arrIterator.next();
-            BFunctionType nextFuncType = m.getType().getAttachedFunctions()[0].type;
-            BRecordType recordType = (BRecordType) ((BUnionType) nextFuncType.retType).getMemberTypes().get(0);
+            Map<String, BField> fields = new HashMap<>();
+            if (arr.getType().getTag() == TypeTags.ARRAY_TAG) {
+                fields.put("value", new BField(arr.getElementType(), "value", Flags.PUBLIC + Flags.REQUIRED));
+            } else {
+                BTupleType tupleType = (BTupleType) arr.getType();
+                LinkedHashSet<BType> types = new LinkedHashSet<>(tupleType.getTupleTypes());
+                if (tupleType.getRestType() != null) {
+                    types.add(tupleType.getRestType());
+                }
+                BType unionType;
+                if (types.size() == 1) {
+                    unionType = types.iterator().next();
+                } else {
+                    unionType = new BUnionType(new ArrayList<>(types));
+                }
+                fields.put("value", new BField(unionType, "value", Flags.PUBLIC + Flags.REQUIRED));
+            }
+
+            BRecordType recordType = new BRecordType("$$returnType$$", null, 0, fields,
+                    null, true, TypeFlags.PURETYPE);
```

Comment:
IIUC, the last parameter `typeFlags` indicates if the record is `anydata`, `pure` (and although irrelevant here, if nillable) right? Since this method can be called with any array/tuple, I guess we have to iterate through the member type(s) and figure out the flags here?

May have to update wherever applicable.